### PR TITLE
Interface reflects support of pagination for media resources loading

### PR DESCRIPTION
**Pull request by boajer**
https://oat-sa.atlassian.net/browse/CON-313

Dependent PRs

- [x] https://github.com/oat-sa/extension-tao-item/pull/445
- [x] https://github.com/oat-sa/extension-tao-mediamanager/pull/245
- [x] https://github.com/oat-sa/tao-core-ui-fe/pull/227 **Need update package.json**

Optional:
- [] https://github.com/oat-sa/extension-tao-devtools/pull/187 tool that should help with testing

### CODE REVIEW

File: models/classes/media/MediaBrowser.php [hunk 1]
```diff
@@ -28,15 +28,19 @@ interface MediaBrowser
...
+     *                'total' => $total
      *               ]
      */
-    public function getDirectory($parentLink = '/', $acceptableMime = [], $depth = 1);
+    public function getDirectory(
+        string $parentLink = '/',
+        array $acceptableMime = [],
+        int $depth = 1,
+        int $limit = 0,
+        int $offset = 0
```
Comment (gabrielfs7) [inline models/classes/media/MediaBrowser.php:42]: Could we have a more explanatory name for these parameters? Cause we return a single directory here and now we have `limit` and `offset`. What is the limit and offset for response of a `single` directory? It is `childrenLimit` and `childrenOffset`? 

Do you think this belongs to this interface?

File: manifest.php [hunk 2]
```diff
@@ -63,7 +63,7 @@
     'label' => 'TAO Base',
     'description' => 'TAO meta-extension',
     'license' => 'GPL-2.0',
-    'version' => '46.8.1',
+    'version' => '46.9.0',
```
Comment (gabrielfs7) [inline manifest.php:66]: This is a breaking change and we should avoid it. I suggest create a new method in another interface and receiving proper parameters and object as response.

**Review by gabrielfs7 — CHANGES_REQUESTED** [review]
We should avoid breaking changes and use deprecation instead. Also some comments about the approach used to solve the problem.

File: models/classes/media/sourceStrategy/HttpSource.php [hunk 3]
```diff
@@ -114,12 +115,18 @@ public function getBaseName($link)
...
      */
-    public function getDirectory($parentLink = '/', $acceptableMime = [], $depth = 1)
+    public function getDirectory(
+        string $parentLink = '',
+        array $acceptableMime = [],
+        int $depth = 1,
+        int $limit = 0,
+        int $offset = 0
+    ): array
     {
```
Comment (shpran) [inline models/classes/media/sourceStrategy/HttpSource.php:128]: [PSR-12](https://www.php-fig.org/psr/psr-2/#44-method-arguments): `When the argument list is split across multiple lines, the closing parenthesis and opening brace MUST be placed together on their own line with one space between them.`
```suggestion
    ): array {
```

**Review by shpran — APPROVED** [review]
1 minor cs comment
Well done 👍🏻 

- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful

**Review by mike-ionut-mihai-sandu-tao — APPROVED** [review]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful

File: models/classes/media/sourceStrategy/HttpSource.php [hunk 4]
```diff
@@ -15,16 +15,18 @@
  * along with this program; if not, write to the Free Software
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  *
- * Copyright (c) 2015 (original work) Open Assessment Technologies SA;
+ * Copyright (c) 2015-2020 (original work) Open Assessment Technologies SA;
```
Comment (bartlomiejmarszal) [inline models/classes/media/sourceStrategy/HttpSource.php:18]: There is no functional changes here. Why you included this in the PR?

File: models/classes/media/sourceStrategy/HttpSource.php [hunk 5]
```diff
@@ -113,13 +115,17 @@ public function getBaseName($link)
         return $realfilename;
     }
 
+    public function getDirectories(QueryObject $params): array
+    {
+        throw new common_Exception('Unable to browse the internet');
```
Comment (bartlomiejmarszal) [inline models/classes/media/sourceStrategy/HttpSource.php:120]: Wait, what?

File: models/classes/media/sourceStrategy/HttpSource.php [hunk 6]
```diff
@@ -113,13 +115,17 @@ public function getBaseName($link)
...
     /**
-     * (non-PHPdoc)
-     * @see \oat\tao\model\media\MediaBrowser::getDirectory()
+     * @inheritDoc
      */
-    public function getDirectory($parentLink = '/', $acceptableMime = [], $depth = 1)
+    public function getDirectory($parentLink = '', $acceptableMime = [], $depth = 1)
     {
-        throw new \common_Exception('Unable to browse the internet');
+        throw new common_Exception('Unable to browse the internet');
```
Comment (bartlomiejmarszal) [inline models/classes/media/sourceStrategy/HttpSource.php:128]: Not following, can you explain me this?

File: models/classes/media/mediaSource/QueryObject.php [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace oat\tao\model\media\MediaSource;
+
+class QueryObject
+{
+    /**
+     * @var string
```
Comment (bartlomiejmarszal) [inline models/classes/media/mediaSource/QueryObject.php:29]: Didn't we agree to use one liner annotations?

**Review by bartlomiejmarszal — CHANGES_REQUESTED** [review]
What happen in HttpSource?

File: models/classes/media/sourceStrategy/HttpSource.php (hunk 4 above)
Comment (boajer) [reply models/classes/media/sourceStrategy/HttpSource.php:18]: This class implements changed interface, so was changed to comply with it

https://github.com/oat-sa/tao-core/pull/2747/files#diff-05ca634d455288c38e2cb9556c48f6f89f6e23c4619ee6f3e668fae082530599R32

File: models/classes/media/sourceStrategy/HttpSource.php (hunk 5 above)
Comment (boajer) [reply models/classes/media/sourceStrategy/HttpSource.php:120]: Could you elaborate, what do you mean? 

New method mimics behavior of the old one, if you have better suggestion about exception text, please  make it, I've kept it same to have same behavior

File: models/classes/media/sourceStrategy/HttpSource.php (hunk 6 above)
Comment (boajer) [reply models/classes/media/sourceStrategy/HttpSource.php:128]: Could you elaborate, what do you mean?

File: models/classes/media/sourceStrategy/HttpSource.php (hunk 5 above)
Comment (bartlomiejmarszal) [reply models/classes/media/sourceStrategy/HttpSource.php:120]: Why `getDirectories` method can only throw an error? 
Also a message in this errors seems odd. Am I missing smoothing obvious?

File: models/classes/media/mediaSource/DirectorySearchQuery.php [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    /** @var int */
+    private $childrenLimit;
+    /** @var int */
+    private $childrenOffset;
+    /** @var MediaAsset */
+    private $asset;
+    /** @var string */
+    private $itemLang;
+    /** @var string */
+    private $itemUri;
```
Comment (gabrielfs7) [inline models/classes/media/mediaSource/DirectorySearchQuery.php:45]: According to our standards:

```suggestion
    /** @var string */
    private $parentLink;
    
    /** @var array */
    private $filter;
    
    /** @var int */
    private $depth;
    
    /** @var int */
    private $childrenLimit;
    
    /** @var int */
    private $childrenOffset;
    
    /** @var MediaAsset */
    private $asset;
    
    /** @var string */
    private $itemLang;
    
    /** @var string */
    private $itemUri;
```

File: test/unit/models/classes/media/mediaSource/QueryObjectTest.php [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+declare(strict_types=1);
+
+namespace oat\tao\model\media\mediaSource;
+
+use oat\generis\test\MockObject;
+use oat\generis\test\TestCase;
+use oat\tao\model\media\MediaAsset;
+
+class QueryObjectTest extends TestCase
```
Comment (gabrielfs7) [inline test/unit/models/classes/media/mediaSource/QueryObjectTest.php:29]: Rename file too, please :) 

```suggestion
class DirectorySearchQueryTest extends TestCase
```

File: test/unit/models/classes/media/mediaSource/QueryObjectTest.php [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        $this->assertSame($this->subject->getParentLink(), 'mediaIdentifier');
+        $this->assertSame($this->subject->getDepth(), 3);
+        $this->assertSame($this->subject->getItemLang(), 'lang');
+        $this->assertSame($this->subject->getItemUri(), 'uri');
+        $this->assertSame($this->subject->getFilter(), ['a' => 'b']);
+        $this->assertSame($this->subject->getChildrenLimit(), 10);
+        $this->assertSame($this->subject->getChildrenOffset(), 11);
+    }
+
```
Comment (gabrielfs7) [inline test/unit/models/classes/media/mediaSource/QueryObjectTest.php:52]:
```suggestion
```

File: models/classes/media/sourceStrategy/HttpSource.php [hunk 11]
```diff
@@ -113,13 +115,17 @@ public function getBaseName($link)
         return $realfilename;
     }
 
+    public function getDirectories(DirectorySearchQuery $params): array
+    {
+        return $this->getDirectory($params->getParentLink());
```
Comment (gabrielfs7) [inline models/classes/media/sourceStrategy/HttpSource.php:120]: Not really required the parameter.

```suggestion
        return $this->getDirectory();
```

File: views/package.json [hunk 12]
```diff
@@ -13,7 +13,7 @@
         "@oat-sa/tao-core-libs": "0.4.2",
         "@oat-sa/tao-core-sdk": "1.7.0",
         "@oat-sa/tao-core-shared-libs": "1.0.3",
-        "@oat-sa/tao-core-ui": "1.19.1",
+        "@oat-sa/tao-core-ui": "git://github.com/oat-sa/tao-core-ui-fe#feature/CON-185/chunk-loading-resource-manager",
```
Comment (gabrielfs7) [inline views/package.json:16]: Please, do not forget to change before merge :)

**Review by mike-ionut-mihai-sandu-tao — COMMENTED** [review]
a few things :
  - the QueryObjectTest need to be called as per the class your testing DirectorySearchQuery
  -  @oat-sa/tao-core-ui need to be updated to version
  - apply psr-12 please
  otherwise good job

**Review by bartlomiejmarszal — CHANGES_REQUESTED** [review]
Just remember about `views/package.json`

File: manifest.php (hunk 2 above)
Comment (gabrielfs7) [inline manifest.php:66]: manifest version is outdate/conflicting. Just a reminder to update when all comments are solved :)

**Review by gabrielfs7 — APPROVED** [review]
Lets have a demo! Good job! :)

**Review by bziondik — APPROVED** [review]
Tested. Works for me.

**Pull request merged**